src/crimson/thread/ThreadPool.cc

```diff
@@ -0,0 +1,77 @@
+#include <chrono>
+
+#include "ThreadPool.h"
+#include "crimson/net/Config.h"
+
+namespace ceph::thread {
+
+ThreadPool::ThreadPool(size_t n)
+  : threads{create_threads(n)}
+{
+  // try to shard the queue evenly
+  const auto queue_size = num_queue_size / seastar::smp::count;
+  for (size_t i = 0; i < seastar::smp::count; i++) {
+    if (i == 0) {
+      free_slots.emplace_back(queue_size + num_queue_size % queue_size);
+    } else {
+      free_slots.emplace_back(queue_size);
+    }
+  }
+}
+
+ThreadPool::~ThreadPool()
+{
+  stop = true;
+  cond.notify_all();
+  for (auto& thread : threads) {
+    thread.join();
+  }
+}
+
+std::vector<std::thread> ThreadPool::create_threads(size_t n)
+{
+  std::vector<std::thread> workers;
+  for (size_t i = 0; i < n; i++) {
+    workers.emplace_back([this] {
+      loop();
+    });
+  }
+  return workers;
+}
+
+bool ThreadPool::have_stray_tasks() const
+{
+  for (auto& free_slot : free_slots) {
+    if (free_slot.waiters() > 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void ThreadPool::loop()
+{
+  using namespace std::chrono_literals;
+  for (;;) {
+    WorkItem* work_item = nullptr;
+    {
+      std::unique_lock lock{mutex};
+      cond.wait_for(lock,
+                    ceph::net::conf.threadpool_empty_queue_max_wait,
+                    [this, &work_item] {
+        return pending.pop(work_item) || stopping();
+      });
+    }
+    if (work_item) {
+      work_item->process();
+    } else if (stopping()) {
+      if (!have_stray_tasks()) {
```

Comment:
the semaphore isn't thread-safe, so i don't think you can reliably read `semaphore::waiters()` here. also, this thread will spin when `stopping() && have_stray_tasks()`

Reply:
neither do i. that's why i added a note in the header. it'd be simpler if we require the user of the pool to follow following protocol when we want to stop it:

1. `gate` the requests sent to the callers of the ThreadPool
2. wait until all the requests are drained.
3. stop the caller
4. stop the thread pool

does it make sense?

Reply:
so probably we don't need to take the stray jobs into consideration, i guess.

Reply:
@cbodley i think a shut-down protocol is the best way to get away from this problem. the reason we need to shard the lockfree queue into multiple subqueue is to let the cores have lock free access the "free_slot". we could have a "total_number_of_waiters" updated by the seastar waiters, and consumed by the alien worker thread. but that's a anti-pattern in seastar.

Reply:
that makes sense, but i think we enforce this protocol part of the interface. maybe ThreadPool should contain something like a `sharded<SubmitQueue>`, where each `SubmitQueue` contains its semaphore and gate, and an async `ThreadPool::shutdown()` waits for this gate on each core before stopping the worker threads?